File: 13-sorting/3-efficient-sorting/main.cpp
// Merge Sort - C++ Example

#include <iostream>
#include <vector>

std::vector<int> merge(const std::vector<int>& left, const std::vector<int>& right) {
    std::vector<int> result;
    int i = 0, j = 0;
    while(i < left.size() && j < right.size()) {
        if(left[i] < right[j]) {
            result.push_back(left[i++]);
        } else {
            result.push_back(right[j++]);
        }
    }
    while(i < left.size()) result.push_back(left[i++]);
    while(j < right.size()) result.push_back(right[j++]);
    return result;
}

std::vector<int> mergeSort(const std::vector<int>& arr) {
    if(arr.size() <=1) return arr;
    int mid = arr.size()/2;
    std::vector<int> left(arr.begin(), arr.begin()+mid);
    std::vector<int> right(arr.begin()+mid, arr.end());
    left = mergeSort(left);
    right = mergeSort(right);
    return merge(left, right);
}

int main() {
    std::vector<int> arr = {38,27,43,3,9,82,10};
    std::vector<int> sorted_arr = mergeSort(arr);

    std::cout << "Sorted array: ";
    for(int x: sorted_arr) std::cout << x << " ";
    std::cout << std::endl;
    return 0;
}

File: 13-sorting/3-efficient-sorting/main.java
// Merge Sort - Java Example

import java.util.Arrays;

class MergeSort {

    public static int[] mergeSort(int[] arr) {
        if (arr.length <= 1) return arr;

        int mid = arr.length / 2;
        int[] left = Arrays.copyOfRange(arr, 0, mid);
        int[] right = Arrays.copyOfRange(arr, mid, arr.length);

        left = mergeSort(left);
        right = mergeSort(right);

        return merge(left, right);
    }

    private static int[] merge(int[] left, int[] right) {
        int[] result = new int[left.length + right.length];
        int i=0, j=0, k=0;

        while(i < left.length && j < right.length) {
            if (left[i] < right[j]) {
                result[k++] = left[i++];
            } else {
                result[k++] = right[j++];
            }
        }
        while(i < left.length) result[k++] = left[i++];
        while(j < right.length) result[k++] = right[j++];

        return result;
    }

    public static void main(String[] args) {
        int[] arr = {38,27,43,3,9,82,10};
        int[] sortedArr = mergeSort(arr);

        System.out.print("Sorted array: ");
        for (int x : sortedArr) System.out.print(x + " ");
        System.out.println();
    }
}
